pyteal/ast/scratchvar.py

```diff
@@ -47,5 +54,86 @@ def load(self) -> ScratchLoad:
         """Load value from Scratch Space"""
         return self.slot.load(self.type)
 
+    def index(self) -> Expr:
+        return self.slot.index()
+
 
 ScratchVar.__module__ = "pyteal"
+
+
+class DynamicScratchVar(ScratchVar):
+    """
+    Example of Dynamic Scratch space whereby the slot index is picked up from the stack:
+        .. code-block:: python1
+
+            player_score = DynamicScratchVar(TealType.uint64)
+
+            wilt = ScratchVar(TealType.uint64, 129)
+            kobe = ScratchVar(TealType.uint64)
+            dt = ScratchVar(TealType.uint64, 131)
+
+            seq = Seq(
+                player_score.set_index(wilt),
+                player_score.store(Int(100)),
+                player_score.set_index(kobe),
+                player_score.store(Int(81)),
+                player_score.set_index(dt),
+                player_score.store(Int(73)),
+                Assert(player_score.load() == Int(73)),
+                Assert(player_score.index() == Int(131)),
+                player_score.set_index(wilt),
+                Assert(player_score.load() == Int(100)),
+                Assert(player_score.index() == Int(129)),
+                Int(100),
+            )
+    """
+
+    def __init__(self, ttype: TealType = TealType.anytype):
+        """Create a new DynamicScratchVar which references other ScratchVar's
+
+        Args:
+            ttype (optional): The type that this variable can hold. Defaults to TealType.anytype.
+        """
+        self.type = ttype
+        self._indexer = ScratchVar(TealType.uint64)
+        self.slot = self._indexer.slot
+
+    def set_index(self, index_var: ScratchVar) -> Expr:
+        """Set this DynamicScratchVar to reference the provided `indexVar`.
+        Followup `store`, `load` and `index` operations will use the provided `indexVar` until
+        `set_index()` is called again to reset the referenced ScratchVar.
+        """
+        if type(index_var) is not ScratchVar:
+            raise TealInputError(
+                "Only allowed to use ScratchVar objects for setting indices, but was given a {}".format(
+                    type(index_var)
+                )
+            )
+
+        return self._indexer.store(index_var.index())
+
+    def storage_type(self) -> TealType:
+        """Get the type of expressions that can be stored in this ScratchVar."""
+        return self.type
+
+    def store(self, value: Expr) -> Expr:
+        """Store the value in the referenced ScratchVar."""
+        require_type(value, self.type)
+        index = ScratchLoad(self._indexer.slot, TealType.uint64)
+        return ScratchStore(slot=None, value=value, index_expression=index)
+
+    def load(self) -> ScratchLoad:
+        """Load the current value from the referenced ScratchVar."""
+        index = ScratchLoad(self._indexer.slot, TealType.uint64)
+        return ScratchLoad(slot=None, type=self.type, index_expression=index)
+
+    def index(self) -> Expr:
```

Comment:
@tzaffi Clarification question:  What motivates the difference in return type for these methods?
* `ScratchVar` returns `ScratchIndex`.
* `DynamicScratchVar.index` returns `ScratchLoad`.

Context:
* As I reviewed, I was surprised to see `index` returns a general type (`Expr`).
* I'm probably missing intuition.  Naively, I'm imagining `ScratchIndex` is the expected return type in all cases.
* As written, I realize `ScratchIndex` requires `ScratchSlot`.  Though it's unclear to me _why_ the `ScratchSlot` dependency exists.
  * There's portions of _IR_ + `ScratchSlot` interaction I still don't follow.  Possibly, the missing intuition lies within _IR_.
  * I _suspect_ the side effects in `ScratchSlot` constructor may also be at play.
  * In my imagination, it's possible to accept either (a) `Int` or (b) union of additional data types.

Reply:
ScratchIndex is a brand new type so I was not super careful here. I think the discrepancy you point out is additional ammunition for @jasonpaulos 's idea to tighten the class hierarchy between ScratchVar and DynamicScratchCar

Reply:
Summarizing a verbal below.  Resolving because the verbal discussion addressed my questions.

* The return types vary because it's not possible to align the types in a way that generates correct TEAL programs.  
* In general, it's a by-design decision to bias towards `Expr` rather than a more specific type to minimize public API changes.